### PR TITLE
test/system: use a random port for the local Docker registry

### PR DESCRIPTION
**Pull request by Rolv-Apneseth**
> Hard-coding the port to 50000 can cause intermittent failures when the port happens to already be in use, and also prevents running the test suite more than once at the same time.
> 
> Let Podman pick a random available port and persist it to a file so test processes in the same run can find it.

See https://github.com/containers/toolbox/pull/1817#issuecomment-4929359777. I _believe_ this should do the trick @debarshiray

### CODE REVIEW

File: test/system/libs/helpers.bash [hunk 1]
```diff
@@ -197,13 +195,25 @@ function _setup_docker_registry() {
...
+  local docker_reg_port
+  docker_reg_port="$(podman --root "${DOCKER_REG_ROOT}" port "${DOCKER_REG_NAME}" 5000)"
+  DOCKER_REG_URI="localhost:${docker_reg_port##*:}"
+  echo "$DOCKER_REG_URI" > "$DOCKER_REG_URI_FILE"
+
+  # Add certificate to Podman's trusted certificates (rootless)
+  run mkdir -p "$HOME"/.config/containers/certs.d/"${DOCKER_REG_URI}"
+  assert_success
+  run cp "${DOCKER_REG_CERTS_DIR}"/domain.crt "$HOME"/.config/containers/certs.d/"${DOCKER_REG_URI}"/domain.crt
+  assert_success
```
[the remark this answers is not in the document]
Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: That's pre-existing code that I just moved, but that makes sense, I could make that change

**Comment by Rolv-Apneseth** [comment]
Rebased after https://github.com/containers/toolbox/pull/1824

File: test/system/libs/helpers.bash (hunk 1 above)
[the remark this answers is not in the document]
Comment (debarshiray) [reply test/system/libs/helpers.bash]: Yes, @Rolv-Apneseth is right.  These uses of `run` already existed in the code.  I took the liberty to remove some of these in https://github.com/containers/toolbox/pull/1825

I couldn't remove all of them because I was running out of time testing all the error paths.  Feel free to submit a pull request removing the rest.  :)

File: test/system/libs/helpers.bash [hunk 2]
```diff
@@ -31,8 +31,12 @@ readonly ROOTLESS_PODMAN_RUNROOT_DIR="$BATS_SUITE_TMPDIR/runroot"
 readonly DOCKER_REG_ROOT="$BATS_SUITE_TMPDIR/docker-registry-root"
 readonly DOCKER_REG_CERTS_DIR="$BATS_SUITE_TMPDIR/certs"
 readonly DOCKER_REG_AUTH_DIR="$BATS_SUITE_TMPDIR/auth"
-readonly DOCKER_REG_URI="localhost:50000"
 readonly DOCKER_REG_NAME="docker-registry"
+# Store the registry URI in a file so that loading this script from different
+# test processes will still use the same URI for the same test suite run. The
+# contents of the file are set when setting up the local Docker registry.
+readonly DOCKER_REG_URI_FILE="$BATS_SUITE_TMPDIR/docker-reg-uri"
+DOCKER_REG_URI="$(cat "$DOCKER_REG_URI_FILE" 2>/dev/null || echo "")"
```
Comment (debarshiray) [inline test/system/libs/helpers.bash]: One quick question: is the `|| echo ""` really needed?

**Review by debarshiray — COMMENTED** [review]
Thanks for working on this, @Rolv-Apneseth !  I had never used `podman port` before - cool.  :)

File: test/system/libs/helpers.bash (hunk 2 above)
Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: I thought so yeah, we want the error to be ignored (when the file hasn't been created yet). Maybe there's a better way?

Comment (debarshiray) [reply test/system/libs/helpers.bash]: Yes, that's what I assumed initially - that you wanted to defend against `set -e`.  Then, when I tried to verify that assumption, I noticed that a missing file doesn't trigger `set -e`.  :O

Reading the Bats [code](https://github.com/bats-core/bats-core/blob/master/lib/bats-core/test_functions.bash), I found that the `source(1)` builtin is invoked as part of the test in an `if` statement where Bash [ignores](https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html) set -e.  I filed https://github.com/bats-core/bats-core/issues/1230 to get a clarification about whether this is coincidence or intentional.

Comment (debarshiray) [reply test/system/libs/helpers.bash]: Assuming that we need to defend against `set -e`, I can't think of a better way.  :)

I could nitpick and say that `true(1)` is slightly better than `echo ""`, because the latter made me wonder if there's something going on with invisible (whitespace?) characters.

**Comment by debarshiray** [comment]
To keep things moving, I took the liberty to rebase this against `main`, and replaced the `|| echo ""` with `|| true`.  Let's see if we hear something from the Bats folks.

**Comment by Rolv-Apneseth** [comment]
Network flake? 

```
not ok 187 run: Smoke test with 'exit 2'
# tags: commands-options
# (from function `assert_success' in file test/system/libs/bats-assert/src/assert.bash, line 114,
#  from function `pull_distro_image' in file test/system/libs/helpers.bash, line 391,
#  from function `pull_default_image' in file test/system/libs/helpers.bash, line 402,
#  from function `create_default_container' in file test/system/libs/helpers.bash, line 470,
#  in test file test/system/104-run.bats, line 806)
#   `create_default_container' failed
# ~ /home/zuul-worker/src/github.com/containers/toolbox
# Failed to load image registry.fedoraproject.org/fedora-toolbox:45 from cache /var/tmp/bats-run-yokdp3/suite/image-cache/fedora-toolbox-45
#
# -- command failed --
# status : 1
# output (3 lines):
#   Getting image source signatures
#   Copying blob sha256:8e0fe36f464ab46d2ce4ea3a1f2779c3921598ebf824e720b8815538ec3d237b
#   time="2026-07-28T00:08:31Z" level=fatal msg="copying config: context deadline exceeded"
# --
```

File: test/system/libs/helpers.bash (hunk 2 above)
Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: Ah, yes, `true` makes more sense, thanks

**Comment by debarshiray** [comment]
recheck

**Comment by debarshiray** [comment]
> Network flake?

It can't be.  The `pull_distro_image()` function does only local disk I/O.  The image is copied from the local image cache directory to the tests' local containers/storage image store with `skopeo copy`.

> ```
> not ok 187 run: Smoke test with 'exit 2'
> # tags: commands-options
> # (from function `assert_success' in file test/system/libs/bats-assert/src/assert.bash, line 114,
> #  from function `pull_distro_image' in file test/system/libs/helpers.bash, line 391,
> #  from function `pull_default_image' in file test/system/libs/helpers.bash, line 402,
> #  from function `create_default_container' in file test/system/libs/helpers.bash, line 470,
> #  in test file test/system/104-run.bats, line 806)
> #   `create_default_container' failed
> # ~ /home/zuul-worker/src/github.com/containers/toolbox
> # Failed to load image registry.fedoraproject.org/fedora-toolbox:45 from cache /var/tmp/bats-run-yokdp3/suite/image-cache/fedora-toolbox-45
> #
> # -- command failed --
> # status : 1
> # output (3 lines):
> #   Getting image source signatures
> #   Copying blob sha256:8e0fe36f464ab46d2ce4ea3a1f2779c3921598ebf824e720b8815538ec3d237b
> #   time="2026-07-28T00:08:31Z" level=fatal msg="copying config: context deadline exceeded"
> # --
> ```

I saw this [once](https://github.com/containers/toolbox/pull/1825#issuecomment-5078361016) recently.  Without looking at the code, the `context deadline exceeded` makes me think that it's a [context.DeadlineExceeded](https://pkg.go.dev/context) inside `skopeo(1)`.  I wonder if we need to increase the 60 second timeout in:
```
skopeo --command-timeout 60s copy ...
```

**Comment by Rolv-Apneseth** [comment]
> > Network flake?
> 
> It can't be.  The `pull_distro_image()` function does only local disk I/O.  The image is copied from the local image cache directory to the tests' local containers/storage image store with `skopeo copy`.
> 
> > ```
> > not ok 187 run: Smoke test with 'exit 2'
> > # tags: commands-options
> > # (from function `assert_success' in file test/system/libs/bats-assert/src/assert.bash, line 114,
> > #  from function `pull_distro_image' in file test/system/libs/helpers.bash, line 391,
> > #  from function `pull_default_image' in file test/system/libs/helpers.bash, line 402,
> > #  from function `create_default_container' in file test/system/libs/helpers.bash, line 470,
> > #  in test file test/system/104-run.bats, line 806)
> > #   `create_default_container' failed
> > # ~ /home/zuul-worker/src/github.com/containers/toolbox
> > # Failed to load image registry.fedoraproject.org/fedora-toolbox:45 from cache /var/tmp/bats-run-yokdp3/suite/image-cache/fedora-toolbox-45
> > #
> > # -- command failed --
> > # status : 1
> > # output (3 lines):
> > #   Getting image source signatures
> > #   Copying blob sha256:8e0fe36f464ab46d2ce4ea3a1f2779c3921598ebf824e720b8815538ec3d237b
> > #   time="2026-07-28T00:08:31Z" level=fatal msg="copying config: context deadline exceeded"
> > # --
> > ```
> 
> I saw this [once](https://github.com/containers/toolbox/pull/1825#issuecomment-5078361016) recently.  Without looking at the code, the `context deadline exceeded` makes me think that it's a [context.DeadlineExceeded](https://pkg.go.dev/context) inside `skopeo(1)`.  I wonder if we need to increase the 60 second timeout in:
> ```
> skopeo --command-timeout 60s copy ...
> ```

Ah, probably a good idea yes

**Review by debarshiray — COMMENTED** [review]
Thanks for bumping the timeouts for local `skopeo copy`!  I merged it through https://github.com/containers/toolbox/pull/1826 so that we can wait a day or two to see if the Bats folks respond to https://github.com/bats-core/bats-core/issues/1230

**Comment by debarshiray** [comment]
I took the liberty to rebase this PR on top of `main`.

**Comment by Rolv-Apneseth** [comment]
```
not ok 1 setup_suite
# (from function `_pull_and_cache_distro_image' in file test/system/libs/helpers.bash, line 139,
#  from function `setup_suite' in test file test/system/setup_suite.bash, line 59)
#   `_pull_and_cache_distro_image ubuntu 20.04' failed
# Failed to cache image quay.io/toolbx/ubuntu-toolbox:20.04 to /var/tmp/bats-run-3FdzJ3/suite/image-cache/ubuntu-toolbox-20.04
# time="2026-07-28T22:40:16Z" level=fatal msg="initializing destination dir:/var/tmp/bats-run-3FdzJ3/suite/image-cache/ubuntu-toolbox-20.04: not a containers image directory, don't want to overwrite important data"
```

**Comment by debarshiray** [comment]
> ```
> not ok 1 setup_suite
> # (from function `_pull_and_cache_distro_image' in file test/system/libs/helpers.bash, line 139,
> #  from function `setup_suite' in test file test/system/setup_suite.bash, line 59)
> #   `_pull_and_cache_distro_image ubuntu 20.04' failed
> # Failed to cache image quay.io/toolbx/ubuntu-toolbox:20.04 to /var/tmp/bats-run-3FdzJ3/suite/image-cache/ubuntu-toolbox-20.04
> # time="2026-07-28T22:40:16Z" level=fatal msg="initializing destination dir:/var/tmp/bats-run-3FdzJ3/suite/image-cache/ubuntu-toolbox-20.04: not a containers image directory, don't want to overwrite important data"
> ```

Oops!  You keep finding some cool corner cases.  It's [ErrNotContainerImageDir](https://pkg.go.dev/go.podman.io/image/v5/directory) from the `go.podman.io/image/v5/directory` package.  :)

File: test/system/libs/helpers.bash [hunk 3]
```diff
@@ -126,6 +130,9 @@ function _pull_and_cache_distro_image() {
       break
     fi
 
+    # Remove stale cache directory left by a failed copy
+    rm -rf "${IMAGE_CACHE_DIR:?}/${image_archive}"
```
Comment (debarshiray) [inline test/system/libs/helpers.bash]: I was just about to hit merge, when I noticed the `:?`.  Is it a stray typo or is the `:?` really needed?

As far as I [know](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html) `${parameter:?word}` is used when `parameter` is null or unset.  In our case `IMAGE_CACHE_DIR` is marked as a `readonly` constant at the top of the file with an almost hard-coded value, and based on the logic above this line, it doesn't look like it can be null or unset.

Or am I missing something?  With shell scripts, I am never sure.  :)

Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: Whoops, you're right, was being overly cautious. I'll remove it.

Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: It was a suggestion by LSP or whatever I have on for bash

Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: Oops, looks like shellcheck actually catches this:

```
In /home/runner/work/toolbox/toolbox/containers/toolbox/builddir/../test/system/libs/helpers.bash line 134:
    rm -rf "${IMAGE_CACHE_DIR}/${image_archive}"
           ^-- SC2115 (warning): Use "${var:?}" to ensure this never expands to / .

For more information:
  https://www.shellcheck.net/wiki/SC2115 -- Use "${var:?}" to ensure this nev...
```

So would you prefer to ignore it or keep the `:?`

Comment (debarshiray) [reply test/system/libs/helpers.bash]: I see - wow!  I never used this trick before.  It certainly looks like a good guideline to adopt, especially given that Steam [bug](https://github.com/ValveSoftware/steam-for-linux/issues/3671).  It will be good if the Toolbx test suite didn't do that.  :)

So, yeah, let's keep it.

Today I learnt.

**Comment by debarshiray** [comment]
> > ```
> > not ok 1 setup_suite
> > # (from function `_pull_and_cache_distro_image' in file test/system/libs/helpers.bash, line 139,
> > #  from function `setup_suite' in test file test/system/setup_suite.bash, line 59)
> > #   `_pull_and_cache_distro_image ubuntu 20.04' failed
> > # Failed to cache image quay.io/toolbx/ubuntu-toolbox:20.04 to /var/tmp/bats-run-3FdzJ3/suite/image-cache/ubuntu-toolbox-20.04
> > # time="2026-07-28T22:40:16Z" level=fatal msg="initializing destination dir:/var/tmp/bats-run-3FdzJ3/suite/image-cache/ubuntu-toolbox-20.04: not a containers image directory, don't want to overwrite important data"
> > ```
> 
> Oops! You keep finding some cool corner cases. It's [ErrNotContainerImageDir](https://pkg.go.dev/go.podman.io/image/v5/directory) from the `go.podman.io/image/v5/directory` package. :)

Merged the patch for this through https://github.com/containers/toolbox/pull/1828

File: test/system/libs/helpers.bash [hunk 4]
```diff
@@ -126,6 +130,9 @@ function _pull_and_cache_distro_image() {
       break
     fi
 
+    # Remove stale cache directory left by a failed copy
```
Comment (debarshiray) [inline test/system/libs/helpers.bash]: Nitpick: this comment is a bit redundant, because we are already explaining it in the Git commit message and the code is obvious.

In comparison, the shell parameter expansion trick in the line below is a lot less obvious.  :)

Comment (debarshiray) [reply test/system/libs/helpers.bash]: > In comparison, the shell parameter expansion trick in the line below is a lot less obvious. :)

I took the liberty to explain it in the Git commit message.

Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: Sounds good, thank you

File: test/system/libs/helpers.bash (hunk 3 above)
Comment (Rolv-Apneseth) [reply test/system/libs/helpers.bash]: Ha, did not know about that Steam bug. Scary

File: test/system/libs/helpers.bash (hunk 2 above)
Comment (debarshiray) [reply test/system/libs/helpers.bash]: Added a note about this to the Git commit message.

**Comment by debarshiray** [comment]
Rebased against `main`.

Let's get this merge after the CI passes.  I will gone on vacation next week, and I don't want this to be forgotten.  :)

**Comment by Rolv-Apneseth** [comment]
> Rebased against `main`.
> 
> Let's get this merge after the CI passes. I will gone on vacation next week, and I don't want this to be forgotten. :)

Nice, me too!
